compose/ui/ui-test-junit4/src/skikoMain/kotlin/androidx/compose/ui/test/ComposeUiTest.skikoMain.kt

```diff
@@ -146,6 +157,14 @@ class SkikoComposeUiTest(
     override fun waitForIdle() {
         // always check even if we are idle
         uncaughtExceptionHandler.throwUncaught()
+        if (!mainClock.autoAdvance) {
+            // Tests with infinite animations should control the Clock manually (via advanceTimeBy).
+            // When infinite animations are present, isIdle will always be false.
```

Comment:
> isIdle will always be false
it's because frameClock.hasAwaiters would always have frame awaiters

Reply:
The correct usage in tests with infinite animations would be:
```kotlin
mainClock.autoAdvanced = false

var state = ....
setContent { UseState(state) }
waitForIdle()

// makes assertions, take a screenshot, etc...

state = newState()
mainClock.advanceTimeByFrame() // or an arbitrary time in ms
waitForIdle()

// makes assertions, take a screenshot, etc...
```

Reply:
It seems, it is only our issue, and Android somehow works in this case? Maybe `hasAwaiters` there returns false in case of infinite animations.

Reply:
Here is what's on android:
```kotlin
override val isIdleNow: Boolean
    get() {
        fun shouldPumpTime(): Boolean {
            hadAwaitersOnMainClock = clock.hasAwaiters
            hadSnapshotChanges = Snapshot.current.hasPendingChanges()
            hadRecomposerChanges = mainRecomposer.hasPendingWork

            val needsRecompose = hadAwaitersOnMainClock || hadSnapshotChanges ||
                hadRecomposerChanges
            return clock.autoAdvance && needsRecompose
        }

        var i = 0
        while (i < 100 && shouldPumpTime()) {
            clock.advanceTimeByFrame()
            ++i
        }

        // pending set content needs all created compose roots,
        // because by definition they will not be in resumed state
        hadPendingSetContent =
            composeRootRegistry.getCreatedComposeRoots().any { it.isBusyAttaching }

        val composeRoots = composeRootRegistry.getRegisteredComposeRoots()
        hadPendingMeasureLayout = composeRoots.any { it.shouldWaitForMeasureAndLayout }

        return !shouldPumpTime() &&
            !hadPendingSetContent &&
            !hadPendingMeasureLayout
    }
```

So `shouldPumpTime` will be always false with manual time management: `return clock.autoAdvance && needsRecompose` regardless of what `hasAwaiters` and other conditions.
I think we can follow this logic too.

Reply:
Yes, let's change it

Reply:
done